python/pdo/contract/response.py

```diff
@@ -137,6 +362,86 @@ def __init__(self, request, response) :
             self.raw_state = self.enclave_service.get_block(state_hash_b64)
             self.new_state_object = ContractState(self.contract_id, self.raw_state)
             self.new_state_object.pull_state_from_eservice(self.enclave_service)
+            
+            # compute ids of blocks in the change set (used for replication)
+            self.new_state_object.compute_ids_of_newblocks(request.contract_state.component_block_ids)
+            self.replicator = Replicator(self.replication_params, self.storage_clients_for_replication, \
+                self.contract_id, self.new_state_object.changed_block_ids, request.enclave_service)
+
+    # -------------------------------------------------------
+    
+    @property
+    def commit_id(self):
+        if self.status and self.state_changed:
+            return (self.contract_id, self.new_state_hash, self.request_number)
+        else:
+            return None
+        
+    # -------------------------------------------------------
+    def replicate_change_set(self):
+        """replicate change set for the current contract state to all provisioned enclaves"""
+            
+        try:
+            self.replicator.replicate_change_set()
+        except ReplicationException as e:
+            logger.exception("Failed to replicate change set %s", str(e))
+            raise Exception("Failed to replicate change set %s", str(e))
+
```

Comment:
So when replication fails we are printing out MANY copies of different messages, sometimes we'll print multiple logger.exception messages. this needs to be made more consistent & less verbose